lib/elixir/pages/anti-patterns/design-anti-patterns.md

```diff
@@ -411,6 +407,94 @@ end
+
+Side-effects are essential when writing production software. Reading a file, logging a message, sending messages between processes, etc. are all side-effects. Writing code with side-effects is not an anti-pattern. However, you must avoid scattering side-effects throughout your functions.
+
+#### Example
+
+Consider a functionality that must parse JSON files from disk, convert all keys to lowercase, and write the updated version to the same file. An initial implementation would look like this:
+
+```elixir
+defmodule Recode do
+  require Logger
+
+  def recode!(file) do
+    data = read_and_decode!(file)
+    data = transform_data(data)
+    encode_and_write!(file, data)
+  end
+
+  defp transform_data(data) do
+    Map.new(data, fn {k, v} -> {String.downcase(k), v} end)
+  end
+
+  defp read_and_decode!(file) do
+    file |> File.read!() |> Jason.decode!()
+  end
+
+  defp encode_and_write!(file, data) do
+    File.write!(file, Jason.encode!(data))
+  end
+end
+```
+
+The code above has side-effects, such as reading and writing files, scattered around different functions. This makes the code harder to understand and maintain. For instance, as the `transform_data/1` function grows in complexity, the only way to test it is by writing and reading data from disk, which makes tests more complex and expensive then necessary.
+
+#### Refactoring
+
+To refactor this anti-pattern, we want break our code into smaller functions without side-effects, and rely on side-effects only when connecting all logic together. To do so, we define a new `recode_contents!` function, which works directly on the contents of a file, instead of file themselves.
+
+```elixir
+defmodule Recode do
+  require Logger
+
+  def recode!(file) do
+    binary = File.read!(file)
+    File.write!(file, recode_contents!(binary))
+  end
+
+  def recode_contents!(binary) when is_binary(binary) do
+    binary
+    |> Jason.decode!()
+    |> transform_data()
+    |> Jason.encode!()
+  end
+
+  defp transform_data(data) do
+    Map.new(data, fn {k, v} -> {String.downcase(k), v} end)
+  end
+end
+```
+
+One of the benefits of unifying the side-effects, alongside with making the `recode_contents!/1` function public, is that it is now possible to test the encoding and decoding logic directly, without having to manipulate files.
+
+Finally, it is worth noting that `recode_contents!` may still raise exceptions when given invalid JSON data and exceptions are also side-effects. Our goal is not to avoid side-effects at all costs, but understand their benefits and impact on our code. After all, exceptions are an important mechanism to signal something went wrong in our code, as long as we are not [*using exceptions for flow control*](#using-exceptions-for-control-flow).
```

Comment:
```suggestion
Finally, it is worth noting that `recode_contents!/1` may still raise exceptions when given invalid JSON data and exceptions are also side-effects. Our goal is not to avoid side-effects at all costs, but understand their benefits and impact on our code. After all, exceptions are an important mechanism to signal something went wrong in our code, as long as we are not [*using exceptions for flow control*](#using-exceptions-for-control-flow).
```